allensdk/brain_observatory/behavior/behavior_project_cache/project_apis/data_io/behavior_ecephys_project_cloud_api.py

```diff
@@ -0,0 +1,336 @@
+
+
+class BehaviorEcephysProjectCloudApi():
+    """API for downloading data released on S3 and returning tables.
+
+    Parameters
+    ----------
+    cache: S3CloudCache
+        an instantiated S3CloudCache object, which has already run
+        `self.load_manifest()` which populates the columns:
+          - metadata_file_names
+          - file_id_column
+    skip_version_check: bool
+        whether to skip the version checking of pipeline SDK version
+        vs. running SDK version, which may raise Exceptions. (default=False)
+    local: bool
+        Whether to operate in local mode, where no data will be downloaded
+        and instead will be loaded from local
+    """
+    def __init__(
+        self,
+        cache: Union[S3CloudCache, LocalCache, StaticLocalCache],
+        skip_version_check: bool = False,
+        local: bool = False
+    ):
+
+        self.cache = cache
+        self.skip_version_check = skip_version_check
+        self._local = local
+        self.load_manifest()
+
+    def load_manifest(self, manifest_name: Optional[str] = None):
+        """
+        Load the specified manifest file into the CloudCache
+
+        Parameters
+        ----------
+        manifest_name: Optional[str]
+            Name of manifest file to load. If None, load latest
+            (default: None)
+        """
+        if manifest_name is None:
+            self.cache.load_last_manifest()
+        else:
+            self.cache.load_manifest(manifest_name)
+
+        expected_metadata = set(["behavior_sessions",
+                                 "sessions",
+                                 "probes",
+                                 "units",
+                                 "channels"])
+
+        if self.cache._manifest.metadata_file_names is None:
+            raise RuntimeError("S3CloudCache object has no metadata "
+                               "file names. BehaviorProjectCloudApi "
+                               "expects a S3CloudCache passed which "
+                               "has already run load_manifest()")
+        cache_metadata = set(self.cache._manifest.metadata_file_names)
+
+        if cache_metadata != expected_metadata:
+            raise RuntimeError("expected S3CloudCache object to have "
+                               f"metadata file names: {expected_metadata} "
+                               f"but it has {cache_metadata}")
+
+        if not self.skip_version_check:
+            data_sdk_version = [i for i in self.cache._manifest._data_pipeline
+                                if i['name'] == "AllenSDK"][0]["version"]
+            version_check(self.cache._manifest.version, data_sdk_version)
+
+        self.logger = logging.getLogger("BehaviorEcephysProjectCloudApi")
+        self._get_ecephys_session_table()
+        self._get_behavior_session_table()
+        self._get_unit_table()
+        self._get_probe_table()
+        self._get_channel_table()
+
+    @classmethod
```

Comment:
If you redefine `from_s3_cloud_cache` in `BehaviorProjectCloudApi` to be a `classmethod` instead of a `staticmethod`, you can do the same trick you did here (calling `cls(cache)` instead of having to explicitly define the class being instantiated). Then, you would not have to duplicate so much code.